examples/research_projects/codeparrot/scripts/codeparrot_training.py

```diff
@@ -203,26 +203,52 @@ def get_lr():
 # Train model
 model.train()
 completed_steps = 0
+config_model = accelerator.unwrap_model(model).config
+t_start = time.time()
 for step, batch in enumerate(train_dataloader, start=1):
     loss = model(batch, labels=batch, use_cache=False).loss
     log_metrics(
         step, {"lr": get_lr(), "samples": step * samples_per_step, "steps": completed_steps, "loss/train": loss.item()}
     )
     loss = loss / args.gradient_accumulation_steps
-    accelerator.backward(loss)
-    if step % args.gradient_accumulation_steps == 0:
+    if step % args.gradient_accumulation_steps != 0:
+        # Prevent backward from doing gradient all_reduce in every step
+        if accelerator.distributed_type == DistributedType.MULTI_GPU:
+            with model.no_sync():
+                accelerator.backward(loss)
+        else:
+            accelerator.backward(loss)
+    else:
+        accelerator.backward(loss)
         accelerator.clip_grad_norm_(model.parameters(), 1.0)
         optimizer.step()
         lr_scheduler.step()
         optimizer.zero_grad()
         completed_steps += 1
+        elapsed_time_per_iteration = time.time() - t_start
+        checkpoint_factor = 4 if args.gradient_checkpointing else 3
+        batch_size = args.train_batch_size * accelerator.state.num_processes * args.gradient_accumulation_steps
+        factor = (
+            24 * checkpoint_factor * batch_size * args.seq_length * config_model.n_layer * (config_model.n_embd**2)
+        )
+        flops_per_iteration = factor * (
+            1.0
+            + (args.seq_length / (6.0 * config_model.n_embd))
+            + (tokenizer.vocab_size / (16.0 * config_model.n_layer * config_model.n_embd))
+        )
+        tflops = flops_per_iteration / (elapsed_time_per_iteration * accelerator.state.num_processes * (10**12))
+        log_metrics(
+            step, {"steps": completed_steps, "tflops": tflops, "time_per_iteration": elapsed_time_per_iteration}
+        )
+        t_start = time.time()
     if step % args.save_checkpoint_steps == 0:
         logger.info("Evaluating and saving model checkpoint")
         eval_loss, perplexity = evaluate(args)
         log_metrics(step, {"loss/eval": eval_loss, "perplexity": perplexity})
         accelerator.wait_for_everyone()
         unwrapped_model = accelerator.unwrap_model(model)
         unwrapped_model.save_pretrained(args.save_dir, save_function=accelerator.save)
+        accelerator.save_state(args.save_dir)
```

Comment:
What exactly does the `save_state` save? A bunch of files? Maybe we could add them to a folder e.g. `args.save_dir + "/state/"`.

Reply:
`save_state` returns a bunch of files (model, optimizer ..), I'm now saving them in folders corresponding to the steps to be able to resume training from these steps later

Reply:
and since `save_state` already saves the model in the folder step, I now use save_pretrained for the unwrapped model only for the last checkpoint to save model in args.save_dir to load direclty from there later